cli: add support for `cockroach dump` to operate on UDTs

rohany opened the pull request:
Fixes #47765.

This PR enables `cockroach dump` to dump databases and tables that
contain user defined types.

Release note (cli change): Allow for `cockroach dump` to dump tables and
databases that contain user defined types.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49808)
<!-- Reviewable:end -->

rohany commented:
Wanted to get a preliminary look at this PR while some open questions around cross database type references are resolved. 

The work done here is maybe another reason why it would be easier to do more of dump on the server side.

knz commented on pkg/cli/dump.go:
```diff
@@ -169,6 +183,17 @@ func runDump(cmd *cobra.Command, args []string) error {
 		}
 
+		if shouldDumpTypes && dumpCtx.dumpMode != dumpDataOnly {
+			if _, err := fmt.Fprintf(w, "SET experimental_enable_enums = true;\n"); err != nil {
```
For cross-version compatibility, be careful to tolerate `dump` talking to a server prior to your changes being available.

knz replied:
(I have commented elsewhere, including on #28948, that the current architecture is flawed - all this logic should be done server-side, in which case  the cross-version compat story is much easier.)

knz commented on pkg/cli/dump.go:
```diff
@@ -260,30 +285,136 @@ type tableMetadata struct {
+		// Parse the create statement into the corresponding AST node. This is done
+		// so that information like the qualified name and enum members can be
+		// parsed out of the definition.
+		stmt, err := parser.ParseOne(createStatement)
```
ok here you're getting a step too far. `dump` should have no business interpreting the strings coming from the server, given that it's likely to run a version behind.

Instead I recommend you operate as follows:

1. make the server able to produce the information you need, via a new `SHOW` statement or `crdb_internal` vtable
2. use it here without interpretaiton

rohany replied:
heh, yeah I felt that I was taking it a step too far here. I'll extend the internal table I'm referencing to return the necessary information.

rohany committed:
cli: add support for `cockroach dump` to operate on UDTs

Fixes #47765.

This PR enables `cockroach dump` to dump databases and tables that
contain user defined types.

Release note (cli change): Allow for `cockroach dump` to dump tables and
databases that contain user defined types.

knz reviewed (commented):
Thanks for updating the PR but this is a bit out of my depth for a quality review right now.
As you hand the review over to another engineer, be mindful to check:
1. that this code works fine when ran with a server running at a lower version
2. that the tests exercise the various error cases

> Reviewed 1 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-:-M9csnuE1E4tiTUZaW3R:b-5jetzv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @rohany)



<!-- Sent from Reviewable.io -->

rohany commented:
@dt can you take a look at this PR. Maybe @jordanlewis could take a look at the enum specific things.

rohany commented:
Pinging for reviews!

rohany commented:
cc @mjibson for some small sqlsmith changes

mjibson commented:
sqlsmith stuff LGTM

dt reviewed (approved):
overall LGTM though like @knz I'd prefer to move some of the more involved enum rendering logic to crdb_internal.create_statements 

> Reviewed 2 of 6 files at r2, 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-:-MAS3Ufl5Wf5Fv8Unspz:bolw5wx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rohany)

___
*[pkg/cli/dump.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-M8qFndFLPiEche-XMlT-r1-187:-MAS2jku1JlknZ7MDG-P:b-ihdp1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf827581da39f265167dee27c7fbcdfda3b42090/pkg/cli/dump.go#L187)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

(I have commented elsewhere, including on #28948, that the current architecture is flawed - all this logic should be done server-side, in which case  the cross-version compat story is much easier.)
</blockquote></details>

this should be fine -- it is _emitting_ a setting that is only known to 20.2+ it isn't sending it to the server. the types are only understood by a new server so it is fine if the setting is too. 
___
*[pkg/cli/dump.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-M8qFndFLPiEche-XMlT-r1-362:-MAS3EfD6Du3E1Tyh3HF:bdrfcqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf827581da39f265167dee27c7fbcdfda3b42090/pkg/cli/dump.go#L362)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

heh, yeah I felt that I was taking it a step too far here. I'll extend the internal table I'm referencing to return the necessary information.
</blockquote></details>

yeah, i'd also have preference for the crdb_internal.create_statements table doing more of this for us than baking a lot of logic into the client.
___
*[pkg/cmd/roachtest/dump.go, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-MAS29Ok81B3SCVMWcdB:-MAS29Ol4KixkE7g8yO6:bxqvv2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb15bb1e03460337cda8665e06f0ac7fc663e509/pkg/cmd/roachtest/dump.go#L1)):*
> ```Go
> // Copyright 2019 The Cockroach Authors.
> ```

nit: 2020


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49808)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/dump.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-M8qFndFLPiEche-XMlT-r1-362:-MAS9nL55kiMirWISl7-:bbyxiny) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf827581da39f265167dee27c7fbcdfda3b42090/pkg/cli/dump.go#L362)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

yeah, i'd also have preference for the crdb_internal.create_statements table doing more of this for us than baking a lot of logic into the client.
</blockquote></details>

I'm not sure what else I can push to the server. Unfortunately, the driver doesn't return arrays as []string, but rather as []byte that needs to be decoded into an array.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49808)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @knz)

___
*[pkg/cli/dump.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49808#-M8qFndFLPiEche-XMlT-r1-362:-MASE8HUDr-vGrcnCxQr:b-5bfytv) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf827581da39f265167dee27c7fbcdfda3b42090/pkg/cli/dump.go#L362)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'm not sure what else I can push to the server. Unfortunately, the driver doesn't return arrays as []string, but rather as []byte that needs to be decoded into an array.
</blockquote></details>

oh, sorry, i was looking at wrong revision in reviewable when i thought that -- this LGTM


<!-- Sent from Reviewable.io -->

rohany commented:
TFTR!

bors r=dt
